src/Npgsql/TypeHandlers/ArrayHandler.cs

```diff
@@ -75,23 +77,40 @@ protected internal override async ValueTask<TRequestedArray> Read<TRequestedArra
         /// <summary>
         /// Reads an array of element type <typeparamref name="TRequestedElement"/> from the given buffer <paramref name="buf"/>.
         /// </summary>
-        protected async ValueTask<Array> ReadArray<TRequestedElement>(NpgsqlReadBuffer buf, bool async, int expectedDimensions = 0)
+        protected async ValueTask<Array> ReadArray<TRequestedElement>(NpgsqlReadBuffer buf, bool async, int expectedDimensions = 0, bool readAsObject = false)
         {
             await buf.Ensure(12, async);
             var dimensions = buf.ReadInt32();
             var containsNulls = buf.ReadInt32() == 1;
             buf.ReadUInt32(); // Element OID. Ignored.
 
-            if (ElementTypeInfo<TRequestedElement>.IsNonNullable && containsNulls)
-                throw new InvalidOperationException(ReadNonNullableCollectionWithNullsExceptionMessage);
+            var returnType = readAsObject && typeof(TRequestedElement).IsValueType
```

Comment:
Doesn't `ElementTypeInfo<T>.IsNonNullable` already cache the same thing that `typeof(TRequestedElement).IsValueType` checks? Not that I'm sure it's necessarily important perf-wise to cache this (does IsValueType already cache internally?), but we're already doing that at the moment.